engine/execution/computation/computer/result_collector.go

```diff
@@ -171,13 +190,32 @@ func (collector *resultCollector) CommitCollection(
 
 func (collector *resultCollector) Stop() {
 	collector.closeOnce.Do(func() {
-		close(collector.viewChan)
-		close(collector.eventsListChan)
+		close(collector.committerInputChan)
+		close(collector.hasherInputChan)
 	})
 }
 
-func (collector *resultCollector) Finalize() *execution.ComputationResult {
+func (collector *resultCollector) Finalize() (
+	*execution.ComputationResult,
+	error,
+) {
 	collector.Stop()
-	collector.wg.Wait()
-	return collector.result
+
+	<-collector.committerDoneChan
+	<-collector.hasherDoneChan
```

Comment:
what about moving the error to the channel?
```
err := <- collector.committerDoneChan

if err != nil {
  ...
}
```

Reply:
it's not safe to return the error in the channel since that could led to a loss event (only the first receiver will see the error, the rest will see nil)